sql/core/src/main/scala/org/apache/spark/sql/execution/BaseScriptTransformationExec.scala

```diff
@@ -329,14 +332,45 @@ case class ScriptTransformationIOSchema(
     schemaLess: Boolean) extends Serializable {
   import ScriptTransformationIOSchema._
 
-  val inputRowFormatMap = inputRowFormat.toMap.withDefault((k) => defaultFormat(k))
-  val outputRowFormatMap = outputRowFormat.toMap.withDefault((k) => defaultFormat(k))
+  val inputRowFormatMap = inputRowFormat.toMap.withDefault(k => defaultFormat(k))
+  val outputRowFormatMap = outputRowFormat.toMap.withDefault(k => defaultFormat(k))
+
+  val separators = (getByte(inputRowFormatMap("TOK_TABLEROWFORMATFIELD"), 0.toByte) ::
+    getByte(inputRowFormatMap("TOK_TABLEROWFORMATCOLLITEMS"), 1.toByte) ::
+    getByte(inputRowFormatMap("TOK_TABLEROWFORMATMAPKEYS"), 2.toByte) :: Nil) ++
+    (4 to 8).map(_.toByte)
```

Comment:
I'm not sure about the current design and why do we need to use different separators for each level?

Reply:
> I'm not sure about the current design and why do we need to use different separators for each level?

Without different seperators for each level, how can we handle data like Array(Array(Array(), Array()))?

Reply:
we cannot write a custom parser for nested arrays just like a json parser?

Reply:
> we cannot write a custom parser for nested arrays just like a json parser?

you mean similar like `DelimitedJSONSerDe` ? I think default way should like `LazySimpleSerde` and keep same with hive.
Construct Json string should be done in spark sql transform(no-serde)'s own serde. @Alfozan have done this and he will raise pr.

Reply:
> why do we need to use different separators for each level?

I have the same question. JSON uses `{}` as the separator and can support arbitrary nested levels.

Reply:
> > why do we need to use different separators for each level?
> 
> I have the same question. JSON uses `{}` as the separator and can support arbitrary nested levels.

IMO, this part should be same with hive for compatibility. If the behavior is not same, migration will have problem in script. And for JSON, we can add a Spark's SerDe.

Reply:
so we must figure out Hive's separator for each level and follow it?

Reply:
> so we must figure out Hive's separator for each level and follow it?

But if user's script is follow this separators, we change this behaviors will cause migration error.

Reply:
I'm still not sure about why the no-serde case needs to follow the `LazySimpleSerde` behaviour.  As for the migration case you suggested above, is it not enough just to use `LazySimpleSerde` in Spark? Most users who care about the hive compatibility just use Spark with `-Phive` enabled, I think.